// slack-web-api-client/dist/client/webhook-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseUrlSender = exports.WebhookSender = void 0;
const errors_1 = require("../errors");
class WebhookSender {
    #url;
    constructor(url) {
        this.#url = url;
    }
    async call(params) {
        try {
            const response = await fetch(this.#url, {
                method: "POST",
                headers: { "Content-Type": "application/json; charset=utf-8" },
                body: JSON.stringify(params),
            });
            const responseBody = await response.text();
            const body = responseBody.toLowerCase();
            if (response.status != 200 || (body !== "ok" && body.toLowerCase() !== '{"ok":true}')) {
                throw new errors_1.WebhookError(response.status, responseBody);
            }
            return response;
        }
        catch (e) {
            throw new errors_1.WebhookError(-1, "", e);
        }
    }
}
exports.WebhookSender = WebhookSender;
exports.ResponseUrlSender = WebhookSender;
