// slack-edge/dist/oauth/error-codes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenIDConnectError = exports.CompletionPageError = exports.InstallationStoreError = exports.InstallationError = exports.MissingCode = exports.InvalidStateParameter = void 0;
/**
 * The state parameter is invalid.
 */
exports.InvalidStateParameter = {
    code: "invalid-state",
    message: "The state parameter is missing or invalid",
};
/**
 * The code parameter is missing.
 */
exports.MissingCode = {
    code: "missing-code",
    message: "The code parameter is missing",
};
/**
 * The app installation failed for some reason.
 */
exports.InstallationError = {
    code: "installation-error",
    message: "The installation process failed",
};
/**
 * The InstallationStore throwed an excetion.
 */
exports.InstallationStoreError = {
    code: "installation-store-error",
    message: "Saving the installation data failed",
};
/**
 * Failed to render the pre-defined completion page.
 */
exports.CompletionPageError = {
    code: "completion-page-failure",
    message: "Rendering the completion page failed",
};
/**
 * Something was wrong with OpenID Connect (Sign in with Slack).
 */
exports.OpenIDConnectError = {
    code: "oidc-error",
    message: "The OpenID Connect process failed",
};
