// slack-web-api-client/dist/client/generated-response/ConversationsOpenResponse.js
"use strict";
// deno-lint-ignore-file ban-unused-ignore no-explicit-any no-empty-interface
///////////////////////////////////
// !!! DO NOT EDIT THIS FILE !!! //
///////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockType = exports.FluffyType = exports.PurpleType = exports.DescriptionType = void 0;
var DescriptionType;
(function (DescriptionType) {
    DescriptionType["Mrkdwn"] = "mrkdwn";
    DescriptionType["PlainText"] = "plain_text";
})(DescriptionType || (exports.DescriptionType = DescriptionType = {}));
var PurpleType;
(function (PurpleType) {
    PurpleType["Broadcast"] = "broadcast";
    PurpleType["Channel"] = "channel";
    PurpleType["Color"] = "color";
    PurpleType["Date"] = "date";
    PurpleType["Emoji"] = "emoji";
    PurpleType["Link"] = "link";
    PurpleType["Team"] = "team";
    PurpleType["Text"] = "text";
    PurpleType["User"] = "user";
    PurpleType["Usergroup"] = "usergroup";
})(PurpleType || (exports.PurpleType = PurpleType = {}));
var FluffyType;
(function (FluffyType) {
    FluffyType["RichTextList"] = "rich_text_list";
    FluffyType["RichTextPreformatted"] = "rich_text_preformatted";
    FluffyType["RichTextQuote"] = "rich_text_quote";
    FluffyType["RichTextSection"] = "rich_text_section";
})(FluffyType || (exports.FluffyType = FluffyType = {}));
var BlockType;
(function (BlockType) {
    BlockType["Actions"] = "actions";
    BlockType["Context"] = "context";
    BlockType["Divider"] = "divider";
    BlockType["Image"] = "image";
    BlockType["RichText"] = "rich_text";
    BlockType["Section"] = "section";
    BlockType["ShareShortcut"] = "share_shortcut";
    BlockType["Video"] = "video";
})(BlockType || (exports.BlockType = BlockType = {}));
